lib/commons/text/is-icon-ligature.js

```diff
@@ -0,0 +1,189 @@
+
+	const style = window.getComputedStyle(textVNode.parent.actualNode);
+	const fontFamily = style.getPropertyValue('font-family');
+
+	if (!fonts[fontFamily]) {
+		fonts[fontFamily] = {
+			occurances: 0,
+			numLigatures: 0
+		};
+	}
+	const font = fonts[fontFamily];
+
+	// improve the performance by only comparing the image data of a font
+	// a certain number of times
+	if (font.occurances >= occuranceThreshold) {
+		// if the font has always been a ligature assume it's a ligature font
+		if (font.numLigatures / font.occurances === 1) {
+			return true;
+		}
+		// inversely, if it's never been a ligature assume it's not a ligature font
+		else if (font.numLigatures === 0) {
+			return false;
+		}
+
+		// we could theoretically get into an odd middle ground scenario in which
+		// the font family is being used as normal text sometimes and as icons
+		// other times. in these cases we would need to always check the text
+		// to know if it's an icon or not
+	}
+	font.occurances++;
+
+	// 30px was chosen to account for common ligatures in normal fonts
+	// such as fi, ff, ffi. If a ligature would add a single column of
+	// pixels to a 30x30 grid, it would not meet the statistical significance
+	// threshold of 10% (30x30 = 900; 30/900 = 3.333%). this also allows for
+	// more than 1 column differences (60/900 = 6.666%) and things like
+	// extending the top of the f in the fi ligature.
+	let fontStyle = `30px ${fontFamily}`;
+
+	// set the size of the canvas to the width of the first letter
+	context.font = fontStyle;
+	const firstChar = nodeValue.charAt(0);
+	let width = context.measureText(firstChar).width;
+
+	// ensure font meets the 30px width requirement (30px font-size doesn't
+	// necessarily mean its 30px wide when drawn)
+	if (width < 30) {
+		const diff = 30 / width;
+		width *= diff;
+		fontStyle = `${30 * diff}px ${fontFamily}`;
+	}
+	canvas.width = canvas.height = width;
+
+	// changing the dimensions of a canvas resets all properties (include font)
+	// and clears it
+	context.font = fontStyle;
+	context.textAlign = 'left';
+	context.textBaseline = 'top';
+	context.fillText(firstChar, 0, 0);
+	const compareData = context.getImageData(0, 0, width, width).data;
+
+	context.clearRect(0, 0, width, width);
+	context.fillText(nodeValue, 0, 0);
+	const compareWith = context.getImageData(0, 0, width, width).data;
+
+	// calculate the number of differences between the first letter and the
+	// entire string
+	let differences = 0;
+	const size = compareWith.length / 4;
+	for (let i = 0; i < compareWith.length; i += 4) {
+		// ignore color differences, only deal with pixel placement differences
+		let compareDataHasPixel = false;
+		let compareWithHasPixel = false;
+		for (let j = 0; j < 4; j++) {
+			compareDataHasPixel = compareDataHasPixel || !!compareData[i + j];
+			compareWithHasPixel = compareWithHasPixel || !!compareWith[i + j];
+		}
+
+		if (compareDataHasPixel !== compareWithHasPixel) {
+			differences++;
```

Comment:
it looks like you can check differences against the pixelThreshold here and break if it exceeds before completing the nested for-loops.

Reply:
we can also shrink the number of iterations if we use a view, combining the 8-bit values into larger ones. In this case, since we're only looking at pixel existence, it seems we should be able to turn these arrays into 32-bit arrays expecting the result to be 0 or non-zero (because we ignore color).

Reply:
The performance hit from this function will be from the two `getImageData` calls and not the double for loop which is just iterating over every pixel once (3600 iterations). Optimizing the loop seems like a micro-optimization when we are already short circuiting the entire algorithm if the font has been seen 3 times.

Reply:
When I looked at it, you're not iterating over each pixel, but each color. I think iterating each pixel would be clearer. Having fewer iterations is a side-effect.

Reply:
something like this:
```js
const compareData = new Uint32Array(context.getImageData(0, 0, width, width).data.buffer);
// ...
const compareWith = new Uint32Array(context.getImageData(0, 0, width, width).data.buffer);
const differences = compareData.reduce((diff, pixel, i) => {
  if (pixel === 0 && compareWith[i] === 0) return diff;
  if (pixel !== 0 && compareWith[i] !== 0) return diff;
  return diff++
}, 0)
```

Reply:
In terms of performance, I don't think converting from an Uint8Array to a Uint32Array (twice) and then iterating over it is significantly faster than looping over a Uint8Array once (I would even counter that it would be slower). 

I also don't think it makes things more clearer as you have to understand the same information to know why you'd want to do the conversion in the first place.